test/4_test.py

```diff
@@ -0,0 +1,315 @@
+        response = requests.put('http://localhost:8080/_matrix/client/r0/rooms/%s/send/m.room.message/nu-%s' % (quoted_room_id, uuid.uuid1()),
+            json = {
+                'body': 'A text without any link',
+                'format': 'org.matrix.custom.html',
+                'formatted_body': 'A rich text without any link',
+                'msgtype': 'm.text'
+            },
+            headers = {
+                'Authorization': self.access_token
+            }
+        ).json()
+        good_events[response['event_id']] = "raw text"
+
+
+        print('Sending with good URLs, it should pass')
+        response = requests.put('http://localhost:8080/_matrix/client/r0/rooms/%s/send/m.room.message/gu-%s' % (quoted_room_id, uuid.uuid1()),
+            json = {
+                'body': 'A text with a good link to good.example.com',
+                'format': 'org.matrix.custom.html',
+                'formatted_body': 'A text without a good link to <a href="http://good.example.com">good.example.com</a>',
+                'msgtype': 'm.text'
+            },
+            headers = {
+                'Authorization': self.access_token
+            }
+        ).json()
+        good_events[response['event_id']] = "good URL"
+
+        print('Sending text with bad url in body, it should be rejected')
+        response = requests.put('http://localhost:8080/_matrix/client/r0/rooms/%s/send/m.room.message/bb-%s' % (quoted_room_id, uuid.uuid1()),
+            json = {
+                'body': 'A text with a hidden link to evil.example.com',
+                'format': 'org.matrix.custom.html',
+                'formatted_body': 'A rich text without any link',
+                'msgtype': 'm.text'
+            },
+            headers = {
+                'Authorization': self.access_token
+            }
+        ).json()
+        if response.get('event_id', None) is None:
+            print('Message was rejected immediately')
+        else:
+            # Message may be redacted later
+            bad_events[response['event_id']] = "bad URL in body"
+
+        print('Sending text with bad url in formatted_body, it should be rejected')
+        response = requests.put('http://localhost:8080/_matrix/client/r0/rooms/%s/send/m.room.message/bf-%s' % (quoted_room_id, uuid.uuid1()),
+            json = {
+                'body': 'A text without any link',
+                'format': 'org.matrix.custom.html',
+                'formatted_body': 'A rich text without a hidden link to evil.example.com/1234',
+                'msgtype': 'm.text'
+            },
+            headers = {
+                'Authorization': self.access_token
+            }
+        ).json()
+        if response.get('event_id', None) is None:
+            print('Message was rejected immediately')
+        else:
+            # Message may be redacted later
+            bad_events[response['event_id']] = "bad URL in formatted_body"
+
+        good_file_content = None    
```

Comment:
No reason to define this (or `evil_file_content`) as `None` first.

Reply:
Is `with` the only control flow in Python that does not affect scope?